### ISSUE TITLE
What's the Caddy mechanism for ensuring that the Middleware instance created with parseCaddyfile will be used for ServeHTTP() rather than my Middleware instances created with New() returned in the ModuleInfo()?

### ISSUE THREAD

**Issue by idavollen** [caddyserver/caddy#6647]
As the Caddy Extension/Plugin docuemntation, https://caddyserver.com/docs/extending-caddy shows, the user defined Middleware could be instantiated from many different places:

1. **New: func() caddy.Module { return new(Middleware) },**
2. **var m Middleware**

```
// CaddyModule returns the Caddy module information.
func (Middleware) CaddyModule() caddy.ModuleInfo {
	return caddy.ModuleInfo{
		ID:  "http.handlers.visitor_ip",
		**New: func() caddy.Module { return new(Middleware) },**
	}
}
```

And 
```
// parseCaddyfile unmarshals tokens from h into a new Middleware.
func parseCaddyfile(h httpcaddyfile.Helper) (caddyhttp.MiddlewareHandler, error) {
	**var m Middleware**
	err := m.UnmarshalCaddyfile(h.Dispenser)
	return m, err
}
```

The **New()** returned in the _ModuleInfo_ for instantiating a new Middleware is totally out of user's control, which means multiple instances of user defined Middleware can be instantiated during module loading, starting process, but without the Caddyfile configurations for the plugin **visitor_ip** being parsed. There exists only _ONE_ instance of Middleware created in the **_parseCaddyfile()_** containing the Caddyfile configs for the visitor_ip. I'm just wondering how Caddy could ensure that the instance of Middleware instantiated with the func parseCaddfile() is being used at runtime for the ServeHTTP()?
```
// ServeHTTP implements caddyhttp.MiddlewareHandler.
func (m Middleware) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyhttp.Handler) error {
	m.w.Write([]byte(r.RemoteAddr))
	return next.ServeHTTP(w, r)
}

func init() {
	caddy.RegisterModule(Middleware{})
	httpcaddyfile.RegisterHandlerDirective("visitor_ip", parseCaddyfile)
}
```
I've been currently developing my local Caddy plugin, and it occurred to me that the instance of my own Middleware instantiated by the **New()** func returned in the ModuleInfo is used by Caddy for the ServeHTTP() rather than the right intance containing the Caddyfile configs for my plugin

**Comment by idavollen** [comment]
@mohammed90 thanks for your prompt feedback on Sunday!

My Caddy plugin is still under development, not published in github yet, however, I tried to print out info on my XAuthMiddleware struct on both the New() returned in ModuleInfo and the parseCaddyfile() fromm caddy run to the ServeHTTP() was firstly called. 

As you can see, the info on my XAuthMiddleware for the first call of ServeHTTP() is **0xc0005eb5e0** that was initially created by the second call of **XAuthMiddleware ModuleInfo.New** &{**0xc0005eb5e0** 0xc000362a80 0xc000337b80 <nil>}
rather than the instance of XAuthMiddleware created by the parsedCaddyfile():
**XAuthMiddleware::parseCaddyfile**;:  &{_**0xc00030e200**_ 0xc000368a80 0xc00039efa0 <nil>}

```
./caddy run --config Caddyfile
XAuthMiddleware::CaddyModule() 
XAuthMiddleware NewXAuth
XAuthMiddleware ModuleInfo.New &{0xc00053e9d0 0xc00036b500 0xc00068c780 <nil>}
2024/10/20 14:01:02.401 INFO    using config from file  {"file": "Caddyfile"}
XAuthMiddleware NewXAuth
XAuthMiddleware::parseCaddyfile;:  &{0xc00030e200 0xc000368a80 0xc00039efa0 <nil>}
after XAuthMiddleware::UnmarshalCaddyfile parseCaddyfile;:  &{0xc00030e200 0xc000368a80 0xc00039efa0 <nil>} web HS256
XAuthMiddleware::CaddyModule() 
2024/10/20 14:01:02.412 INFO    adapted config to JSON  {"adapter": "caddyfile"}
2024/10/20 14:01:02.412 WARN    Caddyfile input is not formatted; run 'caddy fmt --overwrite' to fix inconsistencies    {"adapter": "caddyfile", "file": "Caddyfile", "line": 1}
2024/10/20 14:01:02.427 INFO    admin   admin endpoint started  {"address": "localhost:2019", "enforce_origin": false, "origins": ["//[::1]:2019", "//127.0.0.1:2019", "//localhost:2019"]}
2024/10/20 14:01:02.428 INFO    tls.cache.maintenance   started background certificate maintenance      {"cache": "0xc0005d1380"}
2024/10/20 14:01:02.430 WARN    tls     stapling OCSP   {"error": "no OCSP stapling for [*.ciso.com]: no OCSP server specified in certificate"}
2024/10/20 14:01:02.432 INFO    http.auto_https skipping automatic certificate management because one or more matching certificates are already loaded  {"domain": "traefik.ciso.com", "server_name": "srv0"}
2024/10/20 14:01:02.432 INFO    http.auto_https skipping automatic certificate management because one or more matching certificates are already loaded  {"domain": "secure.ciso.com", "server_name": "srv0"}
2024/10/20 14:01:02.432 INFO    http.auto_https skipping automatic certificate management because one or more matching certificates are already loaded  {"domain": "dao.ciso.com", "server_name": "srv0"}
2024/10/20 14:01:02.433 INFO    http.auto_https enabling automatic HTTP->HTTPS redirects        {"server_name": "srv0"}
XAuthMiddleware NewXAuth
XAuthMiddleware ModuleInfo.New &{0xc0005eb5e0 0xc000362a80 0xc000337b80 <nil>}
XAuthMiddleware::CaddyModule() 
Provision:  &{0xc0005eb5e0 0xc000362a80 0xc000337b80 0xc00069be00}
2024/10/20 14:01:02.440 INFO    http.handlers.xauth     XAuthMiddleware is Provisioning
XAuthMiddleware::Validate:  &{     []} &{0xc0002bb680 false} &{  X-JWT-Token  }
2024/10/20 14:01:02.440 INFO    http.handlers.xauth     XAuthMiddleware is Validating
XAuthMiddleware::Validate   2h0m0s 
2024/10/20 14:01:02.443 WARN    tls     unable to get instance ID; storage clean stamps will be incomplete      {"error": "invalid UUID length: 0"}
2024/10/20 14:01:02.443 INFO    http    enabling HTTP/3 listener        {"addr": ":443"}
2024/10/20 14:01:02.447 INFO    http.log        server running  {"name": "srv0", "protocols": ["h1", "h2", "h3"]}
2024/10/20 14:01:02.447 INFO    http.log        server running  {"name": "remaining_auto_https_redirects", "protocols": ["h1", "h2", "h3"]}
2024/10/20 14:01:02.451 INFO    tls     storage cleaning happened too recently; skipping for now        {"storage": "FileStorage:/Users/dupapa/Library/Application Support/Caddy", "instance": "00000000-0000-0000-0000-000000000000", "try_again": "2024/10/21 14:01:02.451", "try_again_in": 86399.999993109}
2024/10/20 14:01:02.451 INFO    autosaved config (load with --resume flag)      {"file": "/Users/dupapa/Library/Application Support/Caddy/autosave.json"}
2024/10/20 14:01:02.452 INFO    serving initial configuration
2024/10/20 14:01:02.452 INFO    tls     finished cleaning storage units


xauth ServeHTTP is being called /rest/hallo GET &{0xc0005eb5e0 0xc000362a80 0xc000337b80 0xc00069be00}  2h0m0s 
```

**Comment by francislavoie** [comment]
The Caddyfile adapter runs as a separate step, all it does it produce a JSON config, and then that JSON config gets loaded and run. You cannot expect the same struct to be used between adapting and running.

**Comment by idavollen** [comment]
But from the poit view of implementer of a caddy plugin, I should not care about the transition states from Caddyfile adapter to the JSON config, but the ultimate Struct of my Middleware for running the ServeHTTP() should contain the Caddy directives configurations I've specified in the Caddy file for my extended plugin. The current problemm I've encountered is that the ultimate instance of my struct of Middleware does NOT contain the expected config. It seems to me that the internal transition from Caddyfile adapter to the JSON config failed/missed my original configs due to some unknow reasons.

Here it is example of code snippet for my plugin in the Caddyfile:

```
xauth @ap {
		oidc {
			client_id web
			client_secret secret
			redirect_uri https://secure.ciso.com/xauth.ap/callback
			auth_url https://dao.ciso.com/realms/csd-vps/auth
			token_url https://dao.ciso.com/realms/csd-vps/oauth/token
			scope openid profile email
		}
		session {
			cookie_name XAMSession
			secure  true
			http_only true
			same_site Lax
			max_timeout 4
			inactivity_timeout 30
		}
		token {
			sign_key TkZMNSowQmMjOVU2RUB0bm1DJkU3U1VONkd3SGZMbVk=
			sign_alg HS256
			jwt_header X-JWT-Token
			uid X-AUTH-UID
			roles X-AUTH-ROLES
		}
	}
	reverse_proxy http://localhost:5005
```

**Comment by francislavoie** [comment]
Did you use `json` struct tags on the config fields? Did you make them public (uppercase first letter)? That's necessary so that it marshals to JSON properly. You can use `caddy adapt -p` to see how it looks when adapted to JSON. If it's missing anything, then it won't work.

**Comment by idavollen** [comment]
My Caddy plugin is still under development, unstable, not published on github yet.

this is my local Caddy file containing section configs for my plugin **xauth** where it contains three nested configs for **oidc, session, token**. Through debugging, I've found the parseCaddyfile has no problem to parse them and return right key/value for each nested block
```
{
	order xauth before basicauth
}
(cors) {
	@cors_preflight{args[0]} method OPTIONS
	@cors_dao header Origin "https://dao.ciso.com"

	handle @cors_preflight{args[0]} {
		header {
			Access-Control-Allow-Origin "{args[0]}"
			Access-Control-Allow-Methods "GET, POST, PUT, PATCH, DELETE, OPTIONS"
			Access-Control-Allow-Headers "Content-Type, Cache-Control"
			Access-Control-Allow-Credentials true
			Access-Control-Max-Age "3600"
			defer
		}
		respond "" 204
	}

	header {
		Access-Control-Allow-Origin "{args[0]}"
		Access-Control-Expose-Headers *
		Access-Control-Allow-Credentials true
		defer
	}
}

secure.ciso.com {
	@ap path /rest/* /xauth*
	tls ./public.crt ./private.pem
	log
	import cors https://dao.ciso.com
	handle /pub {
		respond "public Resource X"
	}

	xauth @ap {
		oidc {
			client_id web
			client_secret secret
			redirect_uri https://secure.ciso.com/xauth.ap/callback
			auth_url https://dao.ciso.com/realms/csd-vps/auth
			token_url https://dao.ciso.com/realms/csd-vps/oauth/token
			scope openid profile email
		}
		session {
			cookie_name XAMSession
			secure true
			http_only true
			same_site Lax
			max_timeout 4
			inactivity_timeout 30
		}
		token {
			sign_key TkZMNSowQmMjOVU2RUB0bm1DJkU3U1VONkd3SGZMbVk=
			sign_alg HS256
			jwt_header X-JWT-Token
			uid X-AUTH-UID
			roles X-AUTH-ROLES
		}
	}
	reverse_proxy http://localhost:5005
}

dao.ciso.com {
	tls ./public.crt ./private.pem
	log
	reverse_proxy localhost:9998 {
	}
}

traefik.ciso.com {
	tls ./public.crt ./private.pem
	log
	uri strip_prefix /authnz.ap
	reverse_proxy localhost:9999
}

```

I'm attaching the outprint of your suggested command `caddy adapt -p`. As you can see, the converted JSON config only contains my _xauth_ handler, but doesn't have its configs for _oidc, session, token_ due to some unknown reasons

```
 ./caddy adapt -p       
XAuthMiddleware::CaddyModule() 
XAuthMiddleware NewXAuth
XAuthMiddleware ModuleInfo.New &{0xc000342ec0 0xc0002f6540 0xc000374460 <nil>}
2024/10/21 06:36:30.319 INFO    using adjacent Caddyfile
XAuthMiddleware NewXAuth
XAuthMiddleware::parseCaddyfile;:  &{0xc0005fcf00 0xc000310700 0xc00068a320 <nil>}
after XAuthMiddleware::UnmarshalCaddyfile parseCaddyfile;:  &{0xc0005fcf00 0xc000310700 0xc00068a320 <nil>} web HS256
XAuthMiddleware::CaddyModule() 
{
        "apps": {
                "http": {
                        "servers": {
                                "srv0": {
                                        "listen": [
                                                ":443"
                                        ],
                                        "routes": [
                                                {
                                                        "match": [
                                                                {
                                                                        "host": [
                                                                                "traefik.ciso.com"
                                                                        ]
                                                                }
                                                        ],
                                                        "handle": [
                                                                {
                                                                        "handler": "subroute",
                                                                        "routes": [
                                                                                {
                                                                                        "handle": [
                                                                                                {
                                                                                                        "handler": "rewrite",
                                                                                                        "strip_path_prefix": "/authnz.ap"
                                                                                                },
                                                                                                {
                                                                                                        "handler": "reverse_proxy",
                                                                                                        "upstreams": [
                                                                                                                {
                                                                                                                        "dial": "localhost:9999"
                                                                                                                }
                                                                                                        ]
                                                                                                }
                                                                                        ]
                                                                                }
                                                                        ]
                                                                }
                                                        ],
                                                        "terminal": true
                                                },
                                                {
                                                        "match": [
                                                                {
                                                                        "host": [
                                                                                "secure.ciso.com"
                                                                        ]
                                                                }
                                                        ],
                                                        "handle": [
                                                                {
                                                                        "handler": "subroute",
                                                                        "routes": [
                                                                                {
                                                                                        "handle": [
                                                                                                {
                                                                                                        "handler": "headers",
                                                                                                        "response": {
                                                                                                                "deferred": true,
                                                                                                                "set": {
                                                                                                                        "Access-Control-Allow-Credentials": [
                                                                                                                                "true"
                                                                                                                        ],
                                                                                                                        "Access-Control-Allow-Origin": [
                                                                                                                                "https://dao.ciso.com"
                                                                                                                        ],
                                                                                                                        "Access-Control-Expose-Headers": [
                                                                                                                                "*"
                                                                                                                        ]
                                                                                                                }
                                                                                                        }
                                                                                                }
                                                                                        ]
                                                                                },
                                                                                {
                                                                                        "handle": [
                                                                                                {
                                                                                                        "handler": "xauth"
                                                                                                }
                                                                                        ],
                                                                                        "match": [
                                                                                                {
                                                                                                        "path": [
                                                                                                                "/rest/*",
                                                                                                                "/xauth*"
                                                                                                        ]
                                                                                                }
                                                                                        ]
                                                                                },
                                                                                {
                                                                                        "group": "group3",
                                                                                        "handle": [
                                                                                                {
                                                                                                        "handler": "subroute",
                                                                                                        "routes": [
                                                                                                                {
                                                                                                                        "handle": [
                                                                                                                                {
                                                                                                                                        "handler": "headers",
                                                                                                                                        "response": {
                                                                                                                                                "deferred": true,
                                                                                                                                                "set": {
                                                                                                                                                        "Access-Control-Allow-Credentials": [
                                                                                                                                                                "true"
                                                                                                                                                        ],
                                                                                                                                                        "Access-Control-Allow-Headers": [
                                                                                                                                                                "Content-Type, Cache-Control"
                                                                                                                                                        ],
                                                                                                                                                        "Access-Control-Allow-Methods": [
                                                                                                                                                                "GET, POST, PUT, PATCH, DELETE, OPTIONS"
                                                                                                                                                        ],
                                                                                                                                                        "Access-Control-Allow-Origin": [
                                                                                                                                                                "https://dao.ciso.com"
                                                                                                                                                        ],
                                                                                                                                                        "Access-Control-Max-Age": [
                                                                                                                                                                "3600"
                                                                                                                                                        ]
                                                                                                                                                }
                                                                                                                                        }
                                                                                                                                },
                                                                                                                                {
                                                                                                                                        "handler": "static_response",
                                                                                                                                        "status_code": 204
                                                                                                                                }
                                                                                                                        ]
                                                                                                                }
                                                                                                        ]
                                                                                                }
                                                                                        ],
                                                                                        "match": [
                                                                                                {
                                                                                                        "method": [
                                                                                                                "OPTIONS"
                                                                                                        ]
                                                                                                }
                                                                                        ]
                                                                                },
                                                                                {
                                                                                        "group": "group3",
                                                                                        "handle": [
                                                                                                {
                                                                                                        "handler": "subroute",
                                                                                                        "routes": [
                                                                                                                {
                                                                                                                        "handle": [
                                                                                                                                {
                                                                                                                                        "body": "public Resource X",
                                                                                                                                        "handler": "static_response"
                                                                                                                                }
                                                                                                                        ]
                                                                                                                }
                                                                                                        ]
                                                                                                }
                                                                                        ],
                                                                                        "match": [
                                                                                                {
                                                                                                        "path": [
                                                                                                                "/pub"
                                                                                                        ]
                                                                                                }
                                                                                        ]
                                                                                },
                                                                                {
                                                                                        "handle": [
                                                                                                {
                                                                                                        "handler": "reverse_proxy",
                                                                                                        "upstreams": [
                                                                                                                {
                                                                                                                        "dial": "localhost:5005"
                                                                                                                }
                                                                                                        ]
                                                                                                }
                                                                                        ]
                                                                                }
                                                                        ]
                                                                }
                                                        ],
                                                        "terminal": true
                                                },
                                                {
                                                        "match": [
                                                                {
                                                                        "host": [
                                                                                "dao.ciso.com"
                                                                        ]
                                                                }
                                                        ],
                                                        "handle": [
                                                                {
                                                                        "handler": "subroute",
                                                                        "routes": [
                                                                                {
                                                                                        "handle": [
                                                                                                {
                                                                                                        "handler": "reverse_proxy",
                                                                                                        "upstreams": [
                                                                                                                {
                                                                                                                        "dial": "localhost:9998"
                                                                                                                }
                                                                                                        ]
                                                                                                }
                                                                                        ]
                                                                                }
                                                                        ]
                                                                }
                                                        ],
                                                        "terminal": true
                                                }
                                        ],
                                        "tls_connection_policies": [
                                                {
                                                        "match": {
                                                                "sni": [
                                                                        "traefik.ciso.com"
                                                                ]
                                                        },
                                                        "certificate_selection": {
                                                                "any_tag": [
                                                                        "cert0"
                                                                ]
                                                        }
                                                },
                                                {
                                                        "match": {
                                                                "sni": [
                                                                        "secure.ciso.com"
                                                                ]
                                                        },
                                                        "certificate_selection": {
                                                                "any_tag": [
                                                                        "cert0"
                                                                ]
                                                        }
                                                },
                                                {
                                                        "match": {
                                                                "sni": [
                                                                        "dao.ciso.com"
                                                                ]
                                                        },
                                                        "certificate_selection": {
                                                                "any_tag": [
                                                                        "cert0"
                                                                ]
                                                        }
                                                },
                                                {}
                                        ],
                                        "logs": {
                                                "logger_names": {
                                                        "dao.ciso.com": [
                                                                ""
                                                        ],
                                                        "secure.ciso.com": [
                                                                ""
                                                        ],
                                                        "traefik.ciso.com": [
                                                                ""
                                                        ]
                                                }
                                        }
                                }
                        }
                },
                "tls": {
                        "certificates": {
                                "load_files": [
                                        {
                                                "certificate": "./public.crt",
                                                "key": "./private.pem",
                                                "tags": [
                                                        "cert0"
                                                ]
                                        }
                                ]
                        }
                }
        }
}
```

**Comment by mohammed90** [comment]
The object you build in `parseCaddyfile` is marshalled into its JSON representation in this line of `NewRoute`:

https://github.com/caddyserver/caddy/blob/9753c4451077d4459ec10cb3df27ab9dc4456290/caddyconfig/httpcaddyfile/directives.go#L295

Then Caddy unmarshalls the JSON into the object of your module, calls `Provision` on the object, and uses that object moving forward. That's why Francis said your configuration fields need to have at least their first letter uppercased and have `json` tags. I recommend you go through this blog post by the Go team: [JSON and Go](https://go.dev/blog/json).

**Comment by idavollen** [comment]
@mohammed90 @francislavoie  thanks for your help!

My plugin struct has dependency on external Go scs. It's a bit awkward to impose the requirement that the Caddyfile config fields should be exactly one-to-one mapping to my internal struct fields. 
```
type TokenConfig struct {
	SignKey   string
	SignAlg   string
	JWTHeader string
	UID       string
	Roles     string
}

type MySessionManager struct {
	*scs.SessionManager
	IPCheck bool
}

// XAuthMiddleware facilitates JWT (JSON Web Token) authentication.
type XAuthMiddleware struct {
	SessionManager *MySessionManager
	OidcConfig     *oidc.OIDCConfig
	TokenConfig    *TokenConfig
	logger         *zap.Logger
}
```

For the time being, I've come out a workaround, assigning the instantiated XAuthMiddleware instance in the **_parseCaddyfile_** to a global variable __XAuthMiddleware_, and then checking in the New() func for the ModuleInfo if __XAuthMiddleware_ is not nil, if so, just reusing it instead of creating a new instancing by calling _NewXAuth()_ defined as below:

```

func NewXAuth() *XAuthMiddleware {
	fmt.Println("XAuthMiddleware NewXAuth")
	xauth := new(XAuthMiddleware)
	SessionManager := scs.New()
	SessionManager.Lifetime = 2 * time.Hour
	SessionManager.IdleTimeout = 30 * time.Minute
	SessionManager.Cookie.Persist = false
	xauth.SessionManager = &MySessionManager{
		SessionManager: SessionManager,
		IPCheck:        false,
	}
	xauth.OidcConfig = &oidc.OIDCConfig{}
	xauth.TokenConfig = &TokenConfig{
		JWTHeader: "X-JWT-Token",
	}
	return xauth
}
```

Can you shed light of what could be side effects with this workaround?

**Comment by idavollen** [comment]
BTW, since I've decided to take use of the standard caddyfile config adapter, why should I have to additionally ensure that the Caddyfile config should work well with the JSON adapter?

**Comment by francislavoie** [comment]
Without seeing the codebase as a whole, I'm having a really hard time following the logic here.

Your Caddy config structs _should_ have `json` struct tags. See https://caddyserver.com/docs/extending-caddy#configuration which shows that.

> It's a bit awkward to impose the requirement that the Caddyfile config fields should be exactly one-to-one mapping to my internal struct fields.

I don't understand what you mean by this. Why would that "requirement" exist? Your Caddy config struct is your Caddy config struct. There should be no such thing as a "Caddyfile config" struct, your Caddyfile parsing function would just parse the config arguments as text performing any necessary transformation (e.g. duration or byte size strings to plain integers, etc) and then the Caddyfile adapter will serialize it to JSON. But it can only serialize to JSON if you actually declared the `json` struct tags.

> why should I have to additionally ensure that the Caddyfile config should work well with the JSON adapter?

I don't understand that question. The entire point of the Caddyfile adapter is to produce a JSON config. Caddy does not run on Caddyfile, it runs on JSON config which is unmarshaled into Go structs.

**Comment by idavollen** [comment]
I have not managed to find enough documentation yet. The link mentioned by you  https://caddyserver.com/docs/extending-caddy#configuration  only gives a simple use case, what if the json field points to another json object?


https://caddyserver.com/docs/config-adapters says that the caddyfile is the _standard_ adapter. 

The following config adapters are currently available (some are third-party projects):

[caddyfile](https://caddyserver.com/docs/caddyfile) (standard)
[nginx](https://github.com/caddyserver/nginx-adapter)
[jsonc](https://github.com/caddyserver/jsonc-adapter)
[json5](https://github.com/caddyserver/json5-adapter)
[yaml](https://github.com/abiosoft/caddy-yaml)
[cue](https://github.com/caddyserver/cue-adapter)
[toml](https://github.com/awoodbeck/caddy-toml-adapter)
[hcl](https://github.com/francislavoie/caddy-hcl)
[dhall](https://github.com/mholt/dhall-adapter)
[mysql](https://github.com/zhangjiayin/caddy-mysql-adapter)


>I don't understand that question. The entire point of the Caddyfile adapter is to produce a JSON config. Caddy does not run on Caddyfile, it runs on JSON config which is unmarshaled into Go structs.

Since I have already created the instance of my **_XAuthMiddleware_** struct by parsing the Caddyfile configs for my plugin xauth and populating my struct feilds with values specified in the Caddyfile configs, I am confused why this already instantiated **_XAuthMiddleware_** struct can not be reused further by Caddy instead of doing what you mentioned above:

1. Caddyfile adapter produces a JSON config
2. JSON configs is unmarshaled into Go struct, 

If the Go struct at step 2 is my XAuthMiddleware struct, it's already instantiated and populated with values as following


func init() {
	httpcaddyfile.RegisterHandlerDirective("xauth", parseCaddyfile)
}


func parseCaddyfile(h httpcaddyfile.Helper) (caddyhttp.MiddlewareHandler, error) {
	**xauth := NewXAuth()**
	fmt.Println("XAuthMiddleware::parseCaddyfile;: ", xauth)
	err := **xauth.UnmarshalCaddyfile(h.Dispenser)**
	if err != nil {
		fmt.Println("Xauth UnmarshalCaddyfile failed, ", err)
		return nil, err
	}
	return xauth, nil
}

**Comment by francislavoie** [comment]
> what if the json/struct field points to another json object?

Then the child struct should also have JSON tags. The entire config structure of your module needs to be JSON serializable.

> https://caddyserver.com/docs/config-adapters says that the caddyfile is the standard adapter.

Standard means it ships with Caddy by default, as in not provided via a plugin.

You should not have `NewXAuth()` that you call in `parseCaddyfile`. It seems like you're doing that to set defaults. You shouldn't do that, your defaults should be set in your `Provision()` which is run at runtime when loading the config.

The Caddyfile parsing step's job is to take the tokens from the Caddyfile config and apply any that appear to a struct, and leave anything not specified by the user as the zero value (`""` for strings, 0 for numbers, false for booleans, etc). That way all defaults are handled only once in `Provision`, regardless of whether a Caddyfile or JSON or whatever other adapter was used.

**Comment by idavollen** [comment]
>You should not have NewXAuth() that you call in parseCaddyfile. It seems like you're doing that to set defaults. You shouldn't do that, your defaults should be set in your Provision() which is run at runtime when loading the config.

I don't totally agree with you for your suggestion actually makes life harder, IMO
The `NewXAuth()` is called much earlier than `Provision` by the `parseCaddyfile` thanks to `init()`. The body of `NewXAuth()` func sets all default vaules for the struct fields defined in my `XAuthMiddleware`. and then `xauth.UnmarshalCaddyfile()` will overwrite those default vlaues with new values coming from the Caddyfile configs for my plugin/directive **_xauth_** if user doesn't want to use any default values, for instance, user of my plugin can easily override session cookie name as "**user_session**" over the default value _XAMSession_ if he doesn't like the default one. It would be much harder to go through all fields of my `XAuthMiddleware` struct and set default values in `Provision` if a certain struct field has value of nil or empty

**Comment by francislavoie** [comment]
That's wrong. That's not how you should manage config.

- Struct starts with zero values
- UnmarshalCaddyfile only overrides properties that appear in the Caddyfile
- Provision looks for any values that are still zero, and sets them to defaults as necessary.

This approach is the most compatible way, with the least amount of code. It ensures both Caddyfile and JSON configs work and have defaults set properly, and ensures the Caddyfile-adapter-produced-JSON is as small as possible (only having properties that the user actually configured and nothing else).

This is how Caddy modules were designed to be written.

**Comment by idavollen** [comment]
> Struct starts with zero values

for instance, for a Spring boot based Java web app, the session _inactivity_timeout_ by default is _30_ minutes, however, value of 0 means that there is no activity timeout checking for my session, that is, even the user has already been inactive over 30 minutes, her session won't be invalidated, no reauthentication required, which might be desired by some web app, and therefore the user of my plugin xauth explicitly sets inactivity_timeout 0 in the Caddyfile configs for my plugin xauth. How can you handle this well in the Provision func?

I know my plugin better than my uses, and therefore initialize my struct fields with default values meaning for the domain my plugin operates on, providing my users with chances to overwrite those default values with meaningful values in their use cases/scenarios, which is very natural

**Comment by francislavoie** [comment]
> How can you handle this well in the Provision func?

Use `-1` (or any negative number) as "no timeout", and the zero value means "use the default".

> I know my plugin better than my users, and therefore initialize my struct fields with default values meaningful for the domain my plugin operates on, providing my users with chances to overwrite those default values with meaningful values in their use cases/scenarios, which is very natural

Exactly, and you're meant to do that during _provisioning_, so that the defaults actually get set if your users decide to use JSON config directly instead of Caddyfile.

**Comment by idavollen** [comment]
My Caddy plugin struct composites the external SessionManager https://github.com/alexedwards/scs/blob/master/session.go#L16

>Exactly, and you're meant to do that during provisioning, so that the defaults actually get set if your users decide to use JSON config directly instead of Caddyfile.

My way is:
1. provide defaults values in my plugin struct
2. let my users overwrite those default values at will by Caddyfile configs/token (your terminology)

Your currently designed way:
1. let users initialize my struct fields with values from Caddyfile configs/token
2. traverse my struct fields and set those fields whose value is **0, "", false** with default values, which makes it unfeasible for the struct fileds to have **0, "" false** as values set by users, for instance, **_-1_** should be used as you suggested

It should be intuitive which way is natural

BTW, with you caddy plugin design, what is the rationale to create multiple instance of my plugin during its lifetime, especially at module loading intializing time? Is singleton of a caddy plugin better and easier to handle for the host module, Caddy?

**Comment by francislavoie** [comment]
> what is the rationale to create multiple instance of my plugin during its lifetime

Because the Caddyfile adapter may be performed in an entirely different process from the one the Caddy server actually runs in. If you use `caddy reload --config /etc/caddy/Caddyfile` then that runs a CLI process which adapts the Caddyfile (or returns errors if necessary) then makes an HTTP request to the _actually running_ Caddy process at its admin endpoint to push the JSON config to do a zero downtime reload.

Like I said before, separation of concerns. The Caddyfile adapter is entirely separate from runtime.

**Comment by mohammed90** [comment]
To aid you with the design, your struct should have fields for the user configuration, and those are exported with JSON tags (the JSON tags allow for friendly JSON field naming and other options). These act as the configuration message channel between you and the users. It's also helpful to [make the zero values useful](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=385s). nil pointer, empty string, 0, and false can be signals of users wanting the default values. The Caddyfile unmarshaller should act only as an adapter from the more user-friendly format of Caddyfile to the programmatic config format of JSON, because JSON is the native configuration language of Caddy.

In Provision, you're expected to pick up the public/exported fields of the struct (remember, those are user entry points), check whatever default values to be assumed, and translate the public fields to unexported fields (if necessary) that are necessary for the actual operation of the module. For example, the user doesn't need to know of a mutex' presence in the struct, but you might need to initialize one in `Provision` because locking is necessary part of the module operation. So, you have configuration vs. function.

As for the 3rd-party struct, you don't have to expose it directly, but you can (and recommended) abstract it with your own fields in your struct, and then that 3rd-party struct will have an **un**exported instance in your struct that you will fill based on the user's input in your struct.

There's a lot of separation of concern involved. Configuration -> Adapt (if necessary) -> Provision (prepare internal state for operation) -> Validation (e.g. check DB is reachable)  -> Operation (uses internal state set in provision).

**Comment by idavollen** [comment]
https://caddyserver.com/docs/extending-caddy#configuration mentioned by you shows too simple use case. Where can I find the complete documentation on json tags to be used in my struct definition?

It'll be really appreciated if you could provide me with an example of json tags for my struct `XAuthMiddleware` below comprising of nested structs


// XAuthMiddleware Authentication proxy
type XAuthMiddleware struct {
	SessionManager *MySessionManager
	OidcConfig     *OIDCConfig
	TokenConfig    *TokenConfig
	logger         *zap.Logger
}


type TokenConfig struct {
	SignKey   string
	SignAlg   string
	JWTHeader string
	UID       string
	Roles     string
}


type OIDCConfig struct {
	ClientID      string
	ClientSecret  string
	RedirectURI   string
	AuthEndpoint  string
	TokenEndpoint string
	Scope         []string
}

type MySessionManager struct {
	*scs.SessionManager //external dependency on  https://github.com/alexedwards/scs/blob/master/session.go#L16
	IPCheck bool
}

// XAuthMiddleware Authentication proxy
type XAuthMiddleware struct {
	SessionManager *MySessionManager
	OidcConfig     *OIDCConfig
	TokenConfig    *TokenConfig
	logger         *zap.Logger
}

func NewXAuth() *XAuthMiddleware {
	fmt.Println("XAuthMiddleware NewXAuth")
	xauth := new(XAuthMiddleware)
	sessionManager := scs.New()
	sessionManager.Lifetime = 2 * time.Hour
	sessionManager.IdleTimeout = 30 * time.Minute
	sessionManager.Cookie.Persist = false
	xauth.SessionManager = &MySessionManager{
		SessionManager: sessionManager,
		IPCheck:        false,
	}
	xauth.OidcConfig = &OIDCConfig{}
	xauth.TokenConfig = &TokenConfig{
		JWTHeader: "X-JWT-Token",
	}
	return xauth
}


// example of Caddyfile containing token for my plugin
secure.ciso.com {
	@ap path /rest/* /xauth*
	tls ./public.crt ./private.pem
	log
	import cors https://dao.ciso.com
	handle /pub {
		respond "public Resource X"
	}

	xauth @ap {
		oidc {
			client_id web
			client_secret secret
			redirect_uri https://secure.ciso.com/xauth.ap/callback
			auth_endpoint https://dao.ciso.com/realms/csd-vps/auth
			token_endpoint https://dao.ciso.com/realms/csd-vps/oauth/token
			scope openid profile email
		}
		session {
			cookie_name XAMSession
			secure true
			http_only true
			same_site Lax
			max_timeout 4
			inactivity_timeout 30
		}
		token {
			sign_key TkZMNSowQmMjOVU2RUB0bm1DJkU3U1VONkd3SGZMbVk=
			sign_alg HS256
			jwt_header X-JWT-Token
			uid X-AUTH-UID
			roles X-AUTH-ROLES
		}
	}
	reverse_proxy http://localhost:5005
}

**Comment by francislavoie** [comment]
Please make sure to use ```go code blocks when posting.

**Comment by francislavoie** [comment]
Something like this:

```go
type XAuthMiddleware struct {
	SessionConfig *SessionConfig `json:"session,omitempty"`
	OidcConfig    *OIDCConfig    `json:"oidc,omitempty"`
	TokenConfig   *TokenConfig   `json:"token,omitempty"`

	logger         *zap.Logger
	sessionManager *MySessionManager
}

type SessionConfig struct {
	IPCheck bool `json:"ip_check,omitempty"`
}

type TokenConfig struct {
	SignKey   string `json:"sign_key,omitempty"`
	SignAlg   string `json:"sign_alg,omitempty"`
	JWTHeader string `json:"jwt_header,omitempty"`
	UID       string `json:"uid,omitempty"`
	Roles     string `json:"roles,omitempty"`
}

type OIDCConfig struct {
	ClientID      string   `json:"client_id,omitempty"`
	ClientSecret  string   `json:"client_secret,omitempty"`
	RedirectURI   string   `json:"redirect_uri,omitempty"`
	AuthEndpoint  string   `json:"auth_endpoint,omitempty"`
	TokenEndpoint string   `json:"token_endpoint,omitempty"`
	Scopes        []string `json:"scopes,omitempty"`
}

type MySessionManager struct {
	*scs.SessionManager //external dependency on  https://github.com/alexedwards/scs/blob/master/session.go#L16
}
```

Notice that the config structs are _just config_, and the `MySessionManager` should only be initialized during `Provision()` and be fed any config options from `SessionConfig` as necessary.

**Comment by francislavoie** [comment]
Also you should use `caddy.Duration` type for any times, which is a super-type of `time.Duration` (Go stdlib) so then you can parse Caddyfile config like `inactivity_timeout 30s` which automatically multiplies `30` by `seconds`, you could also do `1h` (hour) etc or `500ms`, etc.

**Comment by mohammed90** [comment]
You can also have your own struct for the user-controlled configuration values of `scs.SessionManager`, so you can then translate the config from your own exported struct to the internal `scs.SessionManager` struct which is an implementation detail from the user perspective.

**Comment by idavollen** [comment]
When I woke up, and found your comments, I'm moved, really appreciate your help, bros!!!

**Comment by idavollen** [comment]
By following your way, there are too many boilerplate codes in the `Provision` func for checking each individual struct field has uninitialized value or not

I have to introduce a helper func:
```
func checkValueGeneric(value interface{}) bool {
	v := reflect.ValueOf(value)
	return v.IsValid() && !v.IsZero()
}
```
and go through all fields in all structs, for instance
```
if checkValueGeneric(xauth.SessionConfig.CookieName) {
    sessionManager.Cookie.Name = xauth.SessionConfig.CookieName
} else {
    sessionManager.Cookie.Name = "XAMSession"
}
```

these long boilerplate codes affect code readability and inefficient. I still think my way is more efficeint and the Provision func is just used for provisioning other internal accessible struct fields, such as `logger`

>provide defaults values in my plugin struct
>let my users overwrite those default values at will by Caddyfile configs/token (your terminology)

**Comment by francislavoie** [comment]
Your way is literally wrong because it will not work for users of JSON config. That's what I'm trying to explain. If users used a JSON config they would not get the defaults. This kind of "boilerplate" is just a fact of life in Go. It's not a problem. Just move it to a defaults.go file sibling to do all those conditions and it'll be out of view from your main logic.

**Comment by idavollen** [comment]
I am stuck a bit with your Caddy documentation. https://caddyserver.com/docs/json/ just touches a bit. With JSON config, do you mean that I start Caddy with a JSON file rather than the standard _Caddyfile_?

How can I test JSON config? `./caddy reload --config caddyfile.json` or `./caddy adapt -p`

**Comment by francislavoie** [comment]
To clarify, Caddy's native config language is JSON. That's what Caddy actually runs on. The Caddyfile is there as a UX layer to make it easier for 99% of users to use Caddy and write more natural-reading configs. 

Yes, some users prefer to write JSON config directly, especially if they plan on using Caddy's admin API to perform dynamic/scripted config changes programmatically.

The `caddy adapt` command only runs the Caddyfile adapter and emits the output JSON to stdout. That's it. It's useful to test that your UnmarshalCaddyfile is working correctly (you can have assertions that a given Caddyfile matches an expected JSON output).

The `caddy reload` command only works if you already have Caddy running (i.e. with `caddy run`), it adapts then pushes the JSON config to the running Caddy server over the admin API (the /load endpoint, see https://caddyserver.com/docs/api) which performs a zero downtime in-memory config swap where for a moment two config copies are running at once during the handover. https://caddyserver.com/docs/architecture explains how that works. 

You shouldn't name your JSON config "caddyfile.json" because it's not a Caddyfile, it's a JSON config. That filename is ambiguous.

**Comment by francislavoie** [comment]
Yeah it would run if you just feed Caddy the adapt output, but if a user hand crafted the JSON, the defaults wouldn't be applied because you only did it for the Caddyfile
